crates/sui-core/src/authority_server.rs

```diff
@@ -622,6 +616,160 @@ impl Validator for ValidatorService {
             .get_cache_reader()
             .get_sui_system_state_object_unsafe()?;
 
-        return Ok(tonic::Response::new(response));
+        Ok(tonic::Response::new(response))
+    }
+
+    async fn handle_traffic_req(
+        &self,
+        connection_ip: Option<SocketAddr>,
+        proxy_ip: Option<SocketAddr>,
+    ) -> Result<(), tonic::Status> {
+        if !self.traffic_controller.check(connection_ip, proxy_ip).await {
+            // Entity in blocklist
+            Err(tonic::Status::from_error(SuiError::TooManyRequests.into()))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn handle_traffic_resp<T>(
+        &self,
+        connection_ip: Option<SocketAddr>,
+        proxy_ip: Option<SocketAddr>,
+        response: &Result<tonic::Response<T>, tonic::Status>,
+    ) {
+        let result: SuiResult = if let Err(status) = response {
+            Err(SuiError::from(status.clone()))
+        } else {
+            Ok(())
+        };
+
+        self.traffic_controller.tally(TrafficTally {
+            connection_ip,
+            proxy_ip,
+            result,
+            timestamp: Utc::now(),
```

Comment:
i would prefer we just use SystemTime::now() unless there is a clear reason to pull in a 3rd party package